### DISCUSSION TITLE
curl: (35) OpenSSL/3.3.1: error:80000002:system library::No such file or directory

### DISCUSSION THREAD

**Discussion by danny-su** [openssl/openssl#25172] [Q&A]
I utilized local-ssl-proxy to establish a proxy to the Ollama server, but it failed to forward requests for curl 8.9.1, which depends on OpenSSL 3.3.1.

https://www.npmjs.com/package/local-ssl-proxy

<img width="1006" alt="Warp 2024-08-12 19 37 46" src="https://github.com/user-attachments/assets/8641eab8-bb7e-4c10-977e-3e42bccf1ae8">

**Comment by nhorman** [comment]
that looks like a curl error indicating it can't find the openssl-3.3.1 library.  Did you build curl by yourself?   Whats the output of ldd /path/to/curl ?

**Reply by levitte** [reply]
As a matter of fact, the error line is produced by OpenSSL, so that's not about the OpenSSL library itself.

Proof:

``` console
$ openssl errstr 80000002
error:80000002:system library::No such file or directory
```

**Comment by danny-su** [comment]
@nhorman I installed curl via homebrew, but it works for other SSL sites.

<img width="1017" alt="image" src="https://github.com/user-attachments/assets/bd9a8cc7-c889-4cb5-9245-9dbb39ae5ad4">

**Comment by nhorman** [comment]
You didn't provide the output I requested.  you're probably going to have to take this up with the homebrew packager

**Comment by danny-su** [comment]
@nhorman 

<img width="1012" alt="image" src="https://github.com/user-attachments/assets/44d4f123-6291-4f69-9c4c-21d8f288ecbc">

<img width="1009" alt="image" src="https://github.com/user-attachments/assets/add1cc5d-30ef-40d6-aca6-17dac1b01d9c">

**Comment by nhorman** [comment]
That looks to me like you have a library versioning problem.  Your initial post indicates that your homebrew version of curl is meant to be linked to openssl 3.3.1 while the library that you're resolving at run time points to version 3.0.0.  While they are ABI compatible, the run time loader won't resolve the versioned symbols.  This is really something you need to take up with the apple homebrew community.

**Reply by levitte** [reply]
The library version number 3.0.0 is nonsense.  We make it like this:

https://github.com/openssl/openssl/blob/2a6305dfcd89632b69e49f8b3efe98b7e0daa1aa/Configurations/shared-info.pl#L51

If you look at a produced Makefile, you will see this:

``` console
$ grep SHLIB_VERSION_NUMBER Makefile
SHLIB_VERSION_NUMBER=3
```

So the `.0.0` is something MacOSX adds on its own.

**Comment by tom-cosgrove-arm** [comment]
This looks to me that it should be a discussion - it's not clear that there is an issue with OpenSSL here

**Comment by luckman212** [comment]
This _may_ help:
```
brew reinstall openssl@3 ca-certificates curl
```

**Reply by chen-gloria** [reply]
Worked for me!

**Comment by levitte** [comment]
The error that's produced is an OpenSSL error, so curl does link properly with it.  However, it seems that curl asks OpenSSL to open a file of some sort, and OpenSSL fails to open it, so generates that error.

It's a pity that neither curl nor OpenSSL display the file name.

**Comment by diewellenlaenge** [comment]
I can confirm I have the same issue with OpenSSL 3.3.2 and also 3.3.1.

Fix: use OpenSSL 3.3.0. The error disappeared for me.

> It's a pity that neither curl nor OpenSSL display the file name.

Yes it is. I also used strace (Linux) and procmon (Windows) to find out what file it can't open. Neither of both tracing tools even showed an attempt to open any file. So I think this is a bug since 3.3.1.

**Comment by nhorman** [comment]
can you post your strace output here?

It may be interesting to also run your curl command like this:
```
LD_DEBUG=all curl <options>
```
In the event that its the dynamic loader that is having trouble finding a file, though that would be odd, as that wouldn't result in the OP's report of an openssl error

**Comment by diewellenlaenge** [comment]
I'm afraid I can't, as I use OpenSSL embedded in my application as a static library and it contains too many private strace events, sorry. But I can tell you that the strace output of both versions look the same up to the point where the error from OP appears (3.3.0 = working, 3.3.2 = not working).

**Reply by nhorman** [reply]
Then I'm not sure we can help here.  The reported error is in response to only a few cases without any ancillary data, and those are in response to failure to open a file.  To fully understand the problem we need to know what file is being opened

**Reply by diewellenlaenge** [reply]
I understand. I try to assemble a minimum build soon(ish), one with OpenSSL 3.3.0 and one with 3.3.2.

**Reply by andrewbaxter** [reply]
Could the error not include the file it's trying to open?

**Reply by nhorman** [reply]
openssl does provide the file path in the ancilliary data, I'm not sure why curl doesn't choose to print it out.

**Comment by mlt** [comment]
I experience same issue with static build (at least, didn't try dll) on Windows. The custom app won't work under some environments. It used to work under Visual Studio 2022. Now it won't. It does work with PowerShell and Terminal somehow (disregard another failure. That one is expected). OPENSSL_VERSION_NUMBER = 0x30300010
![image](https://github.com/user-attachments/assets/df774122-9742-40ed-8e4f-83fd021d9923)

**Reply by mlt** [reply]
Turned out (from [procmon](https://learn.microsoft.com/en-us/sysinternals/downloads/procmon)) it was about cert file. I have no idea why it ran earlier as I didn't touch SSL_CERT_FILE.
